In [57]:
import pandas as pd
import numpy as np
import sklearn
from sklearn.metrics import mean_absolute_error as MAE

In [29]:
df = pd.read_csv("train_bm.csv")

In [3]:
#checking rows and columns

df.shape

(8523, 12)

In [4]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8523 entries, 0 to 8522
Data columns (total 12 columns):
 #   Column                     Non-Null Count  Dtype  
---  ------                     --------------  -----  
 0   Item_Identifier            8523 non-null   object 
 1   Item_Weight                7060 non-null   float64
 2   Item_Fat_Content           8523 non-null   object 
 3   Item_Visibility            8523 non-null   float64
 4   Item_Type                  8523 non-null   object 
 5   Item_MRP                   8523 non-null   float64
 6   Outlet_Identifier          8523 non-null   object 
 7   Outlet_Establishment_Year  8523 non-null   int64  
 8   Outlet_Size                6113 non-null   object 
 9   Outlet_Location_Type       8523 non-null   object 
 10  Outlet_Type                8523 non-null   object 
 11  Item_Outlet_Sales          8523 non-null   float64
dtypes: float64(4), int64(1), object(7)
memory usage: 799.2+ KB


In [5]:
# checking number of NULL values

df.isnull().sum()

Item_Identifier                 0
Item_Weight                  1463
Item_Fat_Content                0
Item_Visibility                 0
Item_Type                       0
Item_MRP                        0
Outlet_Identifier               0
Outlet_Establishment_Year       0
Outlet_Size                  2410
Outlet_Location_Type            0
Outlet_Type                     0
Item_Outlet_Sales               0
dtype: int64

### Shuffling and Creating train test model

#### a) Shuffling:


In [6]:
#Playing with the speed of execution (Pandas' SAMPLE vs Scikitlearn's SHUFFLE ) 

%time df.sample(frac=1 , random_state=1)

Wall time: 5 ms


Item_Identifier  Item_Weight Item_Fat_Content  Item_Visibility  \
1070           NCY05        13.50          Low Fat         0.055102   
6305           FDM02        12.50          Regular         0.074035   
8504           NCN18          NaN          Low Fat         0.124111   
5562           FDM02        12.50          Regular         0.073735   
1410           FDL21        15.85          Regular         0.007140   
...              ...          ...              ...              ...   
2895           FDM27          NaN          Regular         0.277459   
7813           FDV38          NaN          Low Fat         0.178193   
905            FDX16        17.85               LF         0.065810   
5192           NCO14         9.60          Low Fat         0.049618   
235            DRM47          NaN          Low Fat         0.043574   

               Item_Type  Item_MRP Outlet_Identifier  \
1070  Health and Hygiene   37.0874            OUT045   
6305              Canned   87.9198            OUT018   
8504           Household  111.7544            OUT027   
5562              Canned   87.1198            OUT046   
1410         Snack Foods   40.8480            OUT013   
...                  ...       ...               ...   
2895                Meat  156.3946            OUT019   
7813               Dairy   54.9956            OUT019   
905         Frozen Foods  149.1050            OUT046   
5192           Household   44.2086            OUT010   
235          Hard Drinks  192.8846            OUT027   

      Outlet_Establishment_Year Outlet_Size Outlet_Location_Type  \
1070                       2002         NaN               Tier 2   
6305                       2009      Medium               Tier 3   
8504                       1985      Medium               Tier 3   
5562                       1997       Small               Tier 1   
1410                       1987        High               Tier 3   
...                         ...         ...                  ...   
2895                       1985       Small               Tier 1   
7813                       1985       Small               Tier 1   
905                        1997       Small               Tier 1   
5192                       1998         NaN               Tier 3   
235                        1985      Medium               Tier 3   

            Outlet_Type  Item_Outlet_Sales  
1070  Supermarket Type1           952.7598  
6305  Supermarket Type2          1133.8574  
8504  Supermarket Type3          4138.6128  
5562  Supermarket Type1          1657.1762  
1410  Supermarket Type1           679.1160  
...                 ...                ...  
2895      Grocery Store           473.3838  
7813      Grocery Store           109.1912  
905   Supermarket Type1          2247.0750  
5192      Grocery Store            44.6086  
235   Supermarket Type3          2293.0152  

[8523 rows x 12 columns]

In [11]:
%time sklearn.utils.shuffle(df , n_samples=df.shape[0])

Wall time: 4.98 ms


Item_Identifier  Item_Weight Item_Fat_Content  Item_Visibility  \
4440           FDQ09        7.235          Low Fat         0.058084   
1622           FDU48       18.850          Low Fat         0.055348   
3993           FDI15          NaN          Low Fat         0.247490   
2905           FDS31       13.100              reg         0.044372   
8154           FDR31          NaN          Regular         0.048925   
...              ...          ...              ...              ...   
8167           DRQ35        9.300          Low Fat         0.042357   
5712           NCC19        6.570          Low Fat         0.096800   
6992           FDF08       14.300          Regular         0.000000   
6577           NCR41       17.850          Low Fat         0.018126   
5570           FDS36        8.380          Regular         0.046848   

                  Item_Type  Item_MRP Outlet_Identifier  \
4440            Snack Foods  113.6834            OUT013   
1622           Baking Goods  131.5284            OUT035   
3993                  Dairy  263.1884            OUT019   
2905  Fruits and Vegetables  180.3318            OUT018   
8154  Fruits and Vegetables  147.4102            OUT027   
...                     ...       ...               ...   
8167            Hard Drinks  123.2388            OUT049   
5712              Household  192.3820            OUT013   
6992  Fruits and Vegetables   89.8856            OUT018   
6577     Health and Hygiene   94.9094            OUT017   
5570           Baking Goods  110.5570            OUT013   

      Outlet_Establishment_Year Outlet_Size Outlet_Location_Type  \
4440                       1987        High               Tier 3   
1622                       2004       Small               Tier 2   
3993                       1985       Small               Tier 1   
2905                       2009      Medium               Tier 3   
8154                       1985      Medium               Tier 3   
...                         ...         ...                  ...   
8167                       1999      Medium               Tier 1   
5712                       1987        High               Tier 3   
6992                       2009      Medium               Tier 3   
6577                       2007         NaN               Tier 2   
5570                       1987        High               Tier 3   

            Outlet_Type  Item_Outlet_Sales  
4440  Supermarket Type1           230.3668  
1622  Supermarket Type1          1581.9408  
3993      Grocery Store           529.9768  
2905  Supermarket Type2          2886.9088  
8154  Supermarket Type3          3207.8244  
...                 ...                ...  
8167  Supermarket Type1          1857.5820  
5712  Supermarket Type1          4247.8040  
6992  Supermarket Type2          1406.1696  
6577  Supermarket Type1           761.6752  
5570  Supermarket Type1          3844.9950  

[8523 rows x 12 columns]

In [34]:
#We'll go with sklearn's shuffle method

new_df = sklearn.utils.shuffle(df ,random_state =1, n_samples=df.shape[0])

In [35]:
new_df.head()

Item_Identifier  Item_Weight Item_Fat_Content  Item_Visibility  \
1070           NCY05        13.50          Low Fat         0.055102   
6305           FDM02        12.50          Regular         0.074035   
8504           NCN18          NaN          Low Fat         0.124111   
5562           FDM02        12.50          Regular         0.073735   
1410           FDL21        15.85          Regular         0.007140   

               Item_Type  Item_MRP Outlet_Identifier  \
1070  Health and Hygiene   37.0874            OUT045   
6305              Canned   87.9198            OUT018   
8504           Household  111.7544            OUT027   
5562              Canned   87.1198            OUT046   
1410         Snack Foods   40.8480            OUT013   

      Outlet_Establishment_Year Outlet_Size Outlet_Location_Type  \
1070                       2002         NaN               Tier 2   
6305                       2009      Medium               Tier 3   
8504                       1985      Medium               Tier 3   
5562                       1997       Small               Tier 1   
1410                       1987        High               Tier 3   

            Outlet_Type  Item_Outlet_Sales  
1070  Supermarket Type1           952.7598  
6305  Supermarket Type2          1133.8574  
8504  Supermarket Type3          4138.6128  
5562  Supermarket Type1          1657.1762  
1410  Supermarket Type1           679.1160

#### b) Spliting

In [42]:
# Creating 4 divisions
div = df.shape[0]//4

# Splitting 3 parts in train and 1 part in test

train_df = new_df.iloc[:3*div+1,:]
test_df = new_df.iloc[3*div+1:,:]

In [46]:
train_df.shape , test_df.shape , new_df.shape , ("final:",train_df.shape[0]+test_df.shape[0])

((6391, 12), (2132, 12), (8523, 12), ('final:', 8523))

### Simple Mean ( mean of Item_Outlet_Sales)

In [50]:
# storing simple mean in a new column in the test set as "simple_mean"
test_df['simple_mean'] = train_df['Item_Outlet_Sales'].mean()

F:\Anaconda3\lib\site-packages\ipykernel_launcher.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  


In [52]:
test_df.head()

Item_Identifier  Item_Weight Item_Fat_Content  Item_Visibility  \
1695           FDR35        12.50          Low Fat         0.020680   
8              FDH17        16.20          Regular         0.016687   
5702           FDZ48        17.75          Low Fat         0.076389   
1066           FDD21        10.30          Regular         0.030563   
1902           FDD29          NaN          Low Fat         0.018321   

                  Item_Type  Item_MRP Outlet_Identifier  \
1695                 Breads  197.2742            OUT013   
8              Frozen Foods   96.9726            OUT045   
5702           Baking Goods  111.1544            OUT017   
1066  Fruits and Vegetables  114.4176            OUT035   
1902           Frozen Foods  255.3698            OUT027   

      Outlet_Establishment_Year Outlet_Size Outlet_Location_Type  \
1695                       1987        High               Tier 3   
8                          2002         NaN               Tier 2   
5702                       2007         NaN               Tier 2   
1066                       2004       Small               Tier 2   
1902                       1985      Medium               Tier 3   

            Outlet_Type  Item_Outlet_Sales  simple_mean  
1695  Supermarket Type1          2587.9646  2184.148305  
8     Supermarket Type1          1076.5986  2184.148305  
5702  Supermarket Type1          1118.5440  2184.148305  
1066  Supermarket Type1          2633.9048  2184.148305  
1902  Supermarket Type3          4819.7262  2184.148305

In [107]:
mae_item_out_sales = MAE(test_df['Item_Outlet_Sales'] , test_df['simple_mean'])
mae_item_out_sales

1351.4516123018486

### Mean `Item Outlet Sales` with respect to `Outlet_Type`

In [81]:
pd.pivot_table(train_df, values = 'Item_Outlet_Sales' , index = 'Outlet_Type' , aggfunc=np.mean)

Item_Outlet_Sales
Outlet_Type                         
Grocery Store             340.212502
Supermarket Type1        2321.512654
Supermarket Type2        2001.738623
Supermarket Type3        3728.741735

In [72]:
#Initializing new column 
test_df['Outlet_type_mean'] = 0

for i in train_df.Outlet_Type.unique():
    test_df['Outlet_type_mean'][test_df['Outlet_Type'] == str(i)] = train_df['Item_Outlet_Sales'][train_df['Outlet_Type'] == str(i)].mean()

F:\Anaconda3\lib\site-packages\ipykernel_launcher.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  
F:\Anaconda3\lib\site-packages\ipykernel_launcher.py:5: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  """
F:\Anaconda3\lib\site-packages\pandas\core\series.py:963: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  self._where(~key, value, inplace=True)
F:\Anaconda3\lib\site-packages

In [108]:
mae_out_type = MAE(test_df['Item_Outlet_Sales'] , test_df['Outlet_type_mean'])
mae_out_type

1145.529062440323

### Mean `Item Outlet Sales` with respect to `Outlet_Establishment_Year`

In [82]:
pd.pivot_table(train_df, values = 'Item_Outlet_Sales' , index = 'Outlet_Establishment_Year' , aggfunc=np.mean)

Item_Outlet_Sales
Outlet_Establishment_Year                   
1985                             2492.716929
1987                             2311.556367
1997                             2251.508728
1998                              346.456688
1999                             2369.278268
2002                             2235.598660
2004                             2432.367907
2007                             2330.067653
2009                             2001.738623

In [95]:
# initializing new column to zero
test_df['Outlet_year_mean'] = 0

for i in train_df.Outlet_Establishment_Year.unique():
    test_df['Outlet_year_mean'][test_df['Outlet_Establishment_Year'] == i] = train_df['Item_Outlet_Sales'][train_df['Outlet_Establishment_Year'] == i].mean()

F:\Anaconda3\lib\site-packages\ipykernel_launcher.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  
F:\Anaconda3\lib\site-packages\ipykernel_launcher.py:5: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  """
F:\Anaconda3\lib\site-packages\pandas\core\series.py:963: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  self._where(~key, value, inplace=True)
F:\Anaconda3\lib\site-packages

In [109]:
mae_out_year = MAE(test_df['Item_Outlet_Sales'] , test_df['Outlet_year_mean'])
mae_out_year

1265.0657298145159

## Mean Item Outlet Sales with respect to Outlet_Location_Type

In [97]:
pd.pivot_table(train_df, values = 'Item_Outlet_Sales', index = ['Outlet_Location_Type'], aggfunc = np.mean)

Item_Outlet_Sales
Outlet_Location_Type                   
Tier 1                      1876.186699
Tier 2                      2332.975458
Tier 3                      2285.026662

In [99]:
# Initializing empty column
test_df['outlet_loc_mean'] = 0

# For every unique entry in Item_Identifier
for i in train_df['Outlet_Location_Type'].unique():
  # calculate and assign mean corresponding to the uniques entries
  test_df['outlet_loc_mean'][test_df['Outlet_Location_Type'] == str(i)] = train_df['Item_Outlet_Sales'][train_df['Outlet_Location_Type'] == str(i)].mean()

F:\Anaconda3\lib\site-packages\ipykernel_launcher.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  
F:\Anaconda3\lib\site-packages\ipykernel_launcher.py:7: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  import sys
F:\Anaconda3\lib\site-packages\pandas\core\series.py:963: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  self._where(~key, value, inplace=True)
F:\Anaconda3\lib\site-p

In [110]:
#calculating mean absolute error
mae_out_loc = MAE(test_df['Item_Outlet_Sales'] , test_df['outlet_loc_mean'] )
mae_out_loc

1345.3413326867874

## Mean Item_Outlet_Sales with respect to both Outlet_Location_Type and Outlet_Establishment_Year

In [103]:
pd.pivot_table(train_df, values = 'Item_Outlet_Sales', index = ['Outlet_Location_Type','Outlet_Establishment_Year'], aggfunc = np.mean)


Item_Outlet_Sales
Outlet_Location_Type Outlet_Establishment_Year                   
Tier 1               1985                              333.545776
                     1997                             2251.508728
                     1999                             2369.278268
Tier 2               2002                             2235.598660
                     2004                             2432.367907
                     2007                             2330.067653
Tier 3               1985                             3728.741735
                     1987                             2311.556367
                     1998                              346.456688
                     2009                             2001.738623

In [104]:
# Initiating new empty column
test_df['Super_mean'] = 0

# Assigning variables to strings ( to shorten code length)
s2 = 'Outlet_Location_Type'
s1 = 'Outlet_Establishment_Year'

# For every Unique Value in s1
for i in test_df[s1].unique():
  # For every Unique Value in s2
  for j in test_df[s2].unique():
    # Calculate and Assign mean to new column, corresponding to both unique values of s1 and s2 simultaneously
    test_df['Super_mean'][(test_df[s1] == i) & (test_df[s2]==str(j))] = train_df['Item_Outlet_Sales'][(train_df[s1] == i) & (train_df[s2]==str(j))].mean()

F:\Anaconda3\lib\site-packages\ipykernel_launcher.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  
F:\Anaconda3\lib\site-packages\ipykernel_launcher.py:13: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  del sys.path[0]
F:\Anaconda3\lib\site-packages\pandas\core\series.py:963: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  self._where(~key, value, inplace=True)
F:\Anaconda3\lib\

In [106]:
#calculating mean absolute error
super_mean_error = MAE(test_df['Item_Outlet_Sales'] , test_df['Super_mean'] )
super_mean_error

1145.126226068051

In [111]:
mae_item_out_sales ,mae_out_type , mae_out_year , mae_out_loc , super_mean_error

(1351.4516123018486,
 1145.529062440323,
 1265.0657298145159,
 1345.3413326867874,
 1145.126226068051)

### Conclusion:
The most better version of the simple mean model has been created with the help of calculaing Mean of `Item_Outlet_Sales` with respect to both `Outlet_Location_Type` and `Outlet_Establishment_Year` which is **1145.126226068051**